### PR TITLE
[MRG] add selector framework to Index, and moltype property to `MinHash` object

### PR DESCRIPTION
**Pull request by ctb**
When working on #809 I decided to try adding a function `Index.select(ksize=ksize, moltype=moltype, ...)` that would return a version of the Index that contained only signatures that matched those requirements. I also added a `moltype` property to the `MinHash` class.

These changes are more general than #809 and touch on a few other issues, like #599, and deserve to be tested and reviewed separately, so I'm breaking them out here.

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

### CODE REVIEW

**Comment by ctb** [comment]
(FWIW I'm thinking of moving the `moltype` property to the `MinHash` object.)

**Comment by ctb** [comment]
I think this is ready for a first round of comments, @luizirber, if you are interested :)

File: sourmash/_minhash.py [hunk 1]
```diff
@@ -566,16 +566,19 @@ def is_molecule_type(self, molecule):
         """Check if this MinHash is a particular human-readable molecule type.
 
         Supports 'protein', 'dayhoff', 'hp', 'DNA'.
+        @CTB deprecate for 4.0?
```
Comment (luizirber) [inline sourmash/_minhash.py:569]: yes. See https://github.com/dib-lab/sourmash/issues/751 for longer discussion, especially https://github.com/dib-lab/sourmash/issues/751#issuecomment-547146368

**Comment by luizirber** [comment]
Do you want to go all the way here and implement `moltype`/`hash_function` as an enum? Like https://github.com/dib-lab/sourmash/commit/a918b261f58959294c958a427cc584106f691954 did (equivalent to #765 in Rust, implementing ideas from #751?)

**Comment by ctb** [comment]
(remaining comments moved to new issues)

**Pull request merged**